File: 15-String/python/04_string_to_integer_atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        s = s.lstrip()
        if not s:
            return 0

        sign = 1
        i = 0
        if s[0] == '-':
            sign = -1
            i = 1
        elif s[0] == '+':
            i = 1

        res = 0
        INT_MAX, INT_MIN = 2**31 - 1, -2**31
        while i < len(s) and s[i].isdigit():
            res = res * 10 + int(s[i])
            if sign == 1 and res > INT_MAX:
                return INT_MAX
            if sign == -1 and -res < INT_MIN:
                return INT_MIN
            i += 1

        return sign * res

File: 15-String/cpp/04-string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string s) {
        int n = s.size();
        if (n == 0) return 0;

        int i = 0;
        while (i < n && s[i] == ' ') i++;
        if (i == n) return 0;

        int sign = 1;
        if (s[i] == '-') {
            sign = -1;
            i++;
        } else if (s[i] == '+') {
            i++;
        }

        long long res = 0;
        while (i < n && isdigit(s[i])) {
            res = res * 10 + (s[i] - '0');
            if (sign == 1 && res > INT_MAX) return INT_MAX;
            if (sign == -1 && -res < INT_MIN) return INT_MIN;
            i++;
        }

        return sign * res;
    }
};
